torch/csrc/autograd/profiler.cpp

```diff
@@ -156,98 +281,126 @@ void enableProfiler(ProfilerConfig config) {
             }
           }
-          pushRange(fn.name(), msg, fn.seqNr(), std::move(inputSizes));
+          state_ptr->pushRange(fn.name(), msg, fn.seqNr(), std::move(inputSizes));
         } else {
-          pushRange(fn.name(), msg, fn.seqNr(), {});
+          state_ptr->pushRange(fn.name(), msg, fn.seqNr(), {});
         }
         return true;
       },
       [](const RecordFunction& fn) {
-        if (fn.getStartCallbacksThreadId() !=
-                RecordFunction::currentThreadId()) {
-          // If we're not in a thread that ran start callbacks, then find
-          // the eventList that was created for the original thread_id. Then,
-          // record the end event on this list so that the block is added to
-          // the correct list, instead of to a new list. This should only run
-          // when calling RecordFunction::end() in a different thread.
-          if (state == ProfilerState::Disabled) {
-            return;
-          } else {
-            std::lock_guard<std::mutex> guard(all_event_lists_map_mutex);
-            const auto& eventListIter =
-                all_event_lists_map.find(fn.getStartCallbacksThreadId());
-            TORCH_INTERNAL_ASSERT(
-                eventListIter != all_event_lists_map.end(),
-                "Did not find thread_id matching ",
-                fn.getStartCallbacksThreadId());
-
-            auto& eventList = eventListIter->second;
-            eventList->record(
-                      EventKind::PopRange,
-                      StringView(""),
-                      fn.getStartCallbacksThreadId(),
-                      state == ProfilerState::CUDA);
-          }
-        } else {
-          popRange();
+        const auto& state = at::ThreadLocalDebugInfo::get(at::DebugInfoKind::PROFILER_STATE);
+        auto state_ptr = dynamic_cast<ProfilerThreadLocalState*>(state.get());
+        if (!state_ptr || state_ptr->config().state == ProfilerState::Disabled) {
+          return;
         }
+        state_ptr->popRange(fn.getStartCallbacksThreadId());
       },
-      /* needs_inputs */ config.report_input_shapes,
-      /* sampling_prob */ 1.0,
-      /* scopes */ {RecordScope::FUNCTION, RecordScope::USER_SCOPE});
-  state = new_state;
-  c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Profiler, true);
+      needs_inputs,
+      /* scopes */ {RecordScope::FUNCTION, RecordScope::USER_SCOPE},
+      CallbackKind::PROFILER);
+}
+
+void removeProfilingCallbacks() {
+  removeCallback(CallbackKind::PROFILER);
+}
 
-  if(state == ProfilerState::CUDA) {
+bool unused_ = []() {
+  at::ThreadLocalState::registerThreadLocalSetting(
+    at::ThreadLocalSetting::PROFILER,
+    []() {
+      auto v = at::SettingValue();
+      v.value = (int64_t)(profiler_nested_depth_ > 0);
+      return v;
+    },
+    [](at::SettingValue v) {
+      // push profiling callbacks in the child task
+      // if profiling is enabled in the parent
+      auto to_push = (bool)v.value;
```

Comment:
maybe I'm missing something, but isn't it broken again?

If profiler is already enabled, then previous version would be `true`. Then you're going to call `setter(true)` twice (at the beginning and at the end) and end up bumping the depth twice.

If you want to preserve getter/setter interface - why not just record the recursive depth as "state". In this situation you'd temporarily carry over the depth from the parent thread to the child thread but then you'd reset it back to whatever the value was on the child thread before picking up the task.

Or you can change these callbacks to be "push/restore"-like as mentioned above. Or you can just ban nested profilers, also good by me :)

Reply:
ah right, yes this won't solve this issue; will upd as per my comment above
push/restore (enter/exit) seems like a solution that might be useful with other settings too, not only profiler